presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java

```diff
@@ -0,0 +1,315 @@
+            context.addSubGraph(graph);
+        }
+        return context.getGraphs();
+    }
+
+    public JoinGraph(PlanNode node)
+    {
+        this(ImmutableList.of(node), ImmutableMultimap.of(), node.getId(), ImmutableList.of(), Optional.empty());
+    }
+
+    public JoinGraph(
+            List<PlanNode> nodes,
+            Multimap<PlanNode, Edge> edges,
+            PlanNodeId rootId,
+            List<Expression> filters,
+            Optional<Map<Symbol, Expression>> assignments)
+    {
+        this.nodes = nodes;
+        this.edges = edges;
+        this.rootId = rootId;
+        this.filters = filters;
+        this.assignments = assignments;
+    }
+
+    public JoinGraph withAssignments(Map<Symbol, Expression> assignments)
+    {
+        return new JoinGraph(nodes, edges, rootId, filters, Optional.of(assignments));
+    }
+
+    public Optional<Map<Symbol, Expression>> getAssignments()
+    {
+        return assignments;
+    }
+
+    public JoinGraph withFilter(Expression expression)
+    {
+        ImmutableList.Builder<Expression> filters = ImmutableList.builder();
+        filters.addAll(this.filters);
+        filters.add(expression);
+
+        return new JoinGraph(nodes, edges, rootId, filters.build(), assignments);
+    }
+
+    public List<Expression> getFilters()
+    {
+        return filters;
+    }
+
+    public PlanNodeId getRootId()
+    {
+        return rootId;
+    }
+
+    public JoinGraph withRootId(PlanNodeId rootId)
+    {
+        return new JoinGraph(nodes, edges, rootId, filters, assignments);
+    }
+
+    public boolean isEmpty()
+    {
+        return nodes.isEmpty();
+    }
+
+    public int size()
+    {
+        return nodes.size();
+    }
+
+    public PlanNode getNode(int index)
+    {
+        return nodes.get(index);
+    }
+
+    public List<PlanNode> getNodes()
+    {
+        return nodes;
+    }
+
+    public Collection<Edge> getEdges(PlanNode node)
+    {
+        return edges.get(node);
+    }
+
+    @Override
+    public String toString()
+    {
+        StringBuilder builder = new StringBuilder();
+
+        for (PlanNode nodeFrom : nodes) {
+            builder.append(nodeFrom.getId())
+                    .append(" = ")
+                    .append(nodeFrom.toString())
+                    .append("\n");
+        }
+        for (PlanNode nodeFrom : nodes) {
+            builder.append(nodeFrom.getId())
+                    .append(":");
+            for (Edge nodeTo : edges.get(nodeFrom)) {
+                builder.append(" ").append(nodeTo.getTargetNode().getId());
+            }
+            builder.append("\n");
+        }
+
+        return builder.toString();
+    }
+
+    private JoinGraph joinWith(JoinGraph other, List<JoinNode.EquiJoinClause> joinClauses, Context context, PlanNodeId rootId)
+    {
+        for (PlanNode node : other.nodes) {
+            checkState(!edges.containsKey(node), format("Node [%s] appeared in two JoinGraphs", node));
```

Comment:
but those two are equivalent. Since I'm effectively (as you pointed out) checking for empty set intersection, whether I iterate over left or right nodes doesn't matter. Those three conditions are basically the same:
```
for each a in A, a is not in B
for each b in B, b is not in A
A intersected with B is empty
```